File: src/MunimMetalkitModule.ts
import { NativeModule, requireNativeModule } from "expo";

import type {
  Animation,
  AnimationDescriptor,
  BinaryData,
  BrushStyle,
  Buffer,
  BufferDescriptor,
  CameraDescriptor,
  Canvas2D,
  Circle2D,
  Color2D,
  ComputeDispatchResult,
  ComputePipelineState,
  DeviceInfo,
  DispatchComputeOptions,
  DrawingLayer,
  Ellipse2D,
  FillStyle,
  LightingDescriptor,
  LineStyle,
  MTLPixelFormat,
  Mesh,
  MeshDescriptor,
  MunimMetalkitModuleEvents,
  Path2D,
  PerformanceInfo,
  Point2D,
  Rectangle2D,
  RenderPipelineState,
  SceneDescriptor,
  Screenshot,
  ScreenshotOptions,
  ShaderDescriptor,
  TextStyle,
  Texture,
  TextureDescriptor,
  TextureLoadOptions,
  TextureRegion,
  TextureTransferOptions,
  ThreadSize,
} from "./MunimMetalkit.types";

/** Marker used in JSDoc: the method rejects with `ERR_NOT_IMPLEMENTED` on every platform. */
type NotImplemented<T> = Promise<T>;

export declare class MunimMetalkitModule extends NativeModule<MunimMetalkitModuleEvents> {
  PI: number;

  // Device

  /** True on iOS devices/simulators with a Metal GPU. Always false on Android and web. */
  isMetalAvailable(): boolean;
  getDeviceInfo(): Promise<DeviceInfo>;

  // Textures

  createTexture(descriptor: TextureDescriptor): Promise<Texture>;
  /** Loads an image from a `file://` or `http(s)://` URL. */
  loadTextureFromURL(
    url: string,
    options?: TextureLoadOptions,
  ): Promise<Texture>;
  /** Decodes encoded image bytes (PNG, JPEG, ...). */
  loadTextureFromData(
    data: BinaryData,
    options?: TextureLoadOptions,
  ): Promise<Texture>;
  /** Uploads tightly packed texels (or `options.bytesPerRow`-strided rows) into a region. */
  updateTexture(
    textureId: string,
    data: BinaryData,
    region?: TextureRegion,
    options?: TextureTransferOptions,
  ): Promise<void>;
  /** Reads texels back. Works for Shared and Private textures. */
  readTexture(
    textureId: string,
    region?: TextureRegion,
    options?: TextureTransferOptions,
  ): Promise<ArrayBuffer>;
  generateMipmaps(textureId: string): Promise<void>;
  releaseTexture(textureId: string): Promise<void>;

  // Buffers

  createBuffer(descriptor: BufferDescriptor): Promise<Buffer>;
  /** `data` must not be empty. */
  createBufferWithData(
    data: BinaryData,
    options?: BufferDescriptor["options"],
  ): Promise<Buffer>;
  updateBuffer(
    bufferId: string,
    data: BinaryData,
    offset?: number,
  ): Promise<void>;
  getBufferContents(bufferId: string): Promise<ArrayBuffer>;
  releaseBuffer(bufferId: string): Promise<void>;

  // Shaders and pipelines

  /**
   * Compiles Metal Shading Language source and returns a library id. Compiler errors reject with
   * `ERR_SHADER_COMPILATION` and include the compiler log.
   */
  createShaderLibrary(source: string): Promise<string>;
  getShaderLibraryFunctionNames(libraryId: string): Promise<string[]>;
  releaseShaderLibrary(libraryId: string): Promise<void>;
  /** Builds a compute pipeline from a `kernel` function in a library. */
  createComputePipelineState(
    libraryId: string,
    functionName: string,
  ): Promise<ComputePipelineState>;
  releaseComputePipelineState(pipelineId: string): Promise<void>;
  /**
   * Encodes one compute dispatch, waits for it to finish, and reports its GPU time.
   * `buffers[i]` is bound at `[[buffer(i)]]`.
   */
  dispatchCompute(
    pipelineId: string,
    buffers: string[],
    threadgroups: ThreadSize,
    options?: DispatchComputeOptions,
  ): Promise<ComputeDispatchResult>;
  /**
   * Builds (and validates) a render pipeline from vertex + fragment functions. There is no JS draw
   * API yet, so the pipeline cannot be used for drawing from JavaScript.
   */
  createRenderPipelineState(
    descriptor: ShaderDescriptor,
  ): Promise<RenderPipelineState>;
  releaseRenderPipelineState(pipelineId: string): Promise<void>;

  // View control: applies to every mounted MunimMetalkitView

  startRendering(): Promise<void>;
  stopRendering(): Promise<void>;
  pauseRendering(): Promise<void>;
  resumeRendering(): Promise<void>;
  /** Renders one frame on every mounted view. */
  setNeedsDisplay(): Promise<void>;
  setPreferredFramesPerSecond(fps: number): Promise<void>;
  setClearColor(
    red: number,
    green: number,
    blue: number,
    alpha: number,
  ): Promise<void>;
  setDrawableSize(width: number, height: number): Promise<void>;

  // Utilities

  /**
   * Captures the most recently mounted MunimMetalkitView as a PNG. Prefer the view ref's
   * `takeScreenshot()` when more than one view is mounted.
   */
  takeScreenshot(options?: ScreenshotOptions): Promise<Screenshot>;
  getPerformanceInfo(): Promise<PerformanceInfo>;

  // Mesh management: not implemented

  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  createMesh(descriptor: MeshDescriptor): NotImplemented<Mesh>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  loadMeshFromURL(url: string): NotImplemented<Mesh>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  loadMeshFromData(
    data: BinaryData,
    format: "obj" | "ply" | "stl",
  ): NotImplemented<Mesh>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  updateMesh(
    meshId: string,
    descriptor: Partial<MeshDescriptor>,
  ): NotImplemented<void>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  releaseMesh(meshId: string): NotImplemented<void>;

  // Animation management: not implemented

  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  createAnimation(descriptor: AnimationDescriptor): NotImplemented<Animation>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  startAnimation(animationId: string): NotImplemented<void>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  pauseAnimation(animationId: string): NotImplemented<void>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  stopAnimation(animationId: string): NotImplemented<void>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setAnimationTime(animationId: string, time: number): NotImplemented<void>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  releaseAnimation(animationId: string): NotImplemented<void>;

  // Scene management: not implemented

  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setScene(scene: SceneDescriptor): NotImplemented<void>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  updateCamera(camera: CameraDescriptor): NotImplemented<void>;
  /** @experimental Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  updateLighting(lighting: LightingDescriptor): NotImplemented<void>;

  // 2D canvas API: declared since 1.x but never had native code. Every call rejects.

  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  createCanvas2D(
    width: number,
    height: number,
    pixelFormat?: MTLPixelFormat,
  ): NotImplemented<Canvas2D>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  clearCanvas2D(
    canvasId: string,
    backgroundColor?: Color2D,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  drawLine2D(
    canvasId: string,
    start: Point2D,
    end: Point2D,
    style: LineStyle,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  drawRectangle2D(
    canvasId: string,
    rect: Rectangle2D,
    style: FillStyle | LineStyle,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  drawCircle2D(
    canvasId: string,
    circle: Circle2D,
    style: FillStyle | LineStyle,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  drawEllipse2D(
    canvasId: string,
    ellipse: Ellipse2D,
    style: FillStyle | LineStyle,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  drawPath2D(
    canvasId: string,
    path: Path2D,
    style: LineStyle,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  drawText2D(
    canvasId: string,
    text: string,
    position: Point2D,
    style: TextStyle,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  measureText2D(
    text: string,
    style: TextStyle,
  ): NotImplemented<{ width: number; height: number }>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  drawImage2D(
    canvasId: string,
    imageTextureId: string,
    destination: Rectangle2D,
    source?: Rectangle2D,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  compositeCanvas2D(
    canvasId: string,
    sourceCanvasId: string,
    operation: string,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  saveCanvas2D(canvasId: string): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  restoreCanvas2D(canvasId: string): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  translateCanvas2D(
    canvasId: string,
    x: number,
    y: number,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  rotateCanvas2D(canvasId: string, angle: number): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  scaleCanvas2D(canvasId: string, x: number, y: number): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setTransformCanvas2D(
    canvasId: string,
    matrix: number[],
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  createDrawingLayer(
    canvasId: string,
    name: string,
  ): NotImplemented<DrawingLayer>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  deleteDrawingLayer(canvasId: string, layerId: string): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setActiveLayer(canvasId: string, layerId: string): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setLayerOpacity(
    canvasId: string,
    layerId: string,
    opacity: number,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setLayerBlendMode(
    canvasId: string,
    layerId: string,
    blendMode: string,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  toggleLayerVisibility(
    canvasId: string,
    layerId: string,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setBrushStyle(canvasId: string, style: BrushStyle): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setLineStyle(canvasId: string, style: LineStyle): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setFillStyle(canvasId: string, style: FillStyle): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setTextStyle(canvasId: string, style: TextStyle): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  resizeCanvas2D(
    canvasId: string,
    width: number,
    height: number,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  cropCanvas2D(canvasId: string, rect: Rectangle2D): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  flipCanvas2D(
    canvasId: string,
    horizontal: boolean,
    vertical: boolean,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  exportCanvas2D(
    canvasId: string,
    format: "png" | "jpg" | "webp",
  ): NotImplemented<ArrayBuffer>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  importImageToCanvas2D(
    canvasId: string,
    imageData: BinaryData,
    position: Point2D,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  getCanvas2DPixel(
    canvasId: string,
    x: number,
    y: number,
  ): NotImplemented<Color2D>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setCanvas2DPixel(
    canvasId: string,
    x: number,
    y: number,
    color: Color2D,
  ): NotImplemented<void>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  getCanvas2DData(canvasId: string): NotImplemented<ArrayBuffer>;
  /** @deprecated Not implemented: rejects with `ERR_NOT_IMPLEMENTED`. */
  setCanvas2DData(canvasId: string, data: BinaryData): NotImplemented<void>;
}

export default requireNativeModule<MunimMetalkitModule>("MunimMetalkit");

File: android/src/main/java/expo/modules/munimmetalkit/MunimMetalkitModule.kt
package expo.modules.munimmetalkit

import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Metal is Apple-only, so on Android the module is an honest stub: `isMetalAvailable()` returns
 * false and every async method rejects with ERR_METAL_UNAVAILABLE. (1.x shipped an unregistered
 * OpenGL ES renderer and a createTexture that never had a GL context; both were removed.)
 */
class MunimMetalkitModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("MunimMetalkit")

    Constant("PI") { Math.PI }

    Function("isMetalAvailable") { false }

    // Arity per method, so calls with the documented arguments reach the rejection below
    // instead of failing on an argument-count mismatch.
    for ((name, arity) in ASYNC_METHODS) {
      when (arity) {
        0 -> AsyncFunction(name) { -> throw MetalUnavailableException(name) }
        1 -> AsyncFunction(name) { _: Any? -> throw MetalUnavailableException(name) }
        2 -> AsyncFunction(name) { _: Any?, _: Any? -> throw MetalUnavailableException(name) }
        3 -> AsyncFunction(name) { _: Any?, _: Any?, _: Any? -> throw MetalUnavailableException(name) }
        else -> AsyncFunction(name) { _: Any?, _: Any?, _: Any?, _: Any? -> throw MetalUnavailableException(name) }
      }
    }

    View(MunimMetalkitView::class) {
      Events("onLoad", "onRender", "onError")
    }
  }

  companion object {
    /** Keep in sync with src/MunimMetalkitModule.ts. */
    private val ASYNC_METHODS = listOf(
      "getDeviceInfo" to 0,
      "createTexture" to 1,
      "loadTextureFromURL" to 2,
      "loadTextureFromData" to 2,
      "updateTexture" to 4,
      "readTexture" to 3,
      "generateMipmaps" to 1,
      "releaseTexture" to 1,
      "createBuffer" to 1,
      "createBufferWithData" to 2,
      "updateBuffer" to 3,
      "getBufferContents" to 1,
      "releaseBuffer" to 1,
      "createShaderLibrary" to 1,
      "getShaderLibraryFunctionNames" to 1,
      "releaseShaderLibrary" to 1,
      "createComputePipelineState" to 2,
      "releaseComputePipelineState" to 1,
      "dispatchCompute" to 4,
      "createRenderPipelineState" to 1,
      "releaseRenderPipelineState" to 1,
      "startRendering" to 0,
      "stopRendering" to 0,
      "pauseRendering" to 0,
      "resumeRendering" to 0,
      "setNeedsDisplay" to 0,
      "setPreferredFramesPerSecond" to 1,
      "setClearColor" to 4,
      "setDrawableSize" to 2,
      "takeScreenshot" to 1,
      "getPerformanceInfo" to 0,
      "createMesh" to 1,
      "loadMeshFromURL" to 1,
      "loadMeshFromData" to 2,
      "updateMesh" to 2,
      "releaseMesh" to 1,
      "createAnimation" to 1,
      "startAnimation" to 1,
      "pauseAnimation" to 1,
      "stopAnimation" to 1,
      "setAnimationTime" to 2,
      "releaseAnimation" to 1,
      "setScene" to 1,
      "updateCamera" to 1,
      "updateLighting" to 1,
      "createCanvas2D" to 3,
      "clearCanvas2D" to 2,
      "drawLine2D" to 4,
      "drawRectangle2D" to 3,
      "drawCircle2D" to 3,
      "drawEllipse2D" to 3,
      "drawPath2D" to 3,
      "drawText2D" to 4,
      "measureText2D" to 2,
      "drawImage2D" to 4,
      "compositeCanvas2D" to 3,
      "saveCanvas2D" to 1,
      "restoreCanvas2D" to 1,
      "translateCanvas2D" to 3,
      "rotateCanvas2D" to 2,
      "scaleCanvas2D" to 3,
      "setTransformCanvas2D" to 2,
      "createDrawingLayer" to 2,
      "deleteDrawingLayer" to 2,
      "setActiveLayer" to 2,
      "setLayerOpacity" to 3,
      "setLayerBlendMode" to 3,
      "toggleLayerVisibility" to 2,
      "setBrushStyle" to 2,
      "setLineStyle" to 2,
      "setFillStyle" to 2,
      "setTextStyle" to 2,
      "resizeCanvas2D" to 3,
      "cropCanvas2D" to 2,
      "flipCanvas2D" to 3,
      "exportCanvas2D" to 2,
      "importImageToCanvas2D" to 3,
      "getCanvas2DPixel" to 3,
      "setCanvas2DPixel" to 4,
      "getCanvas2DData" to 1,
      "setCanvas2DData" to 2,
    )
  }
}

class MetalUnavailableException(method: String) :
  CodedException(
    "ERR_METAL_UNAVAILABLE",
    "munim-metalkit: $method() is unavailable on Android. Metal is only available on iOS.",
    null
  )
